src/error-handling/converting-error-types.md

```diff
@@ -19,15 +19,28 @@ The `From::from` call here means we attempt to convert the error type to the
 type returned by the function:
 
 ```rust,editable
-use std::{fs, io};
-use std::io::Read;
+use std::error::Error;
+use std::fmt::{self, Display, Formatter};
+use std::fs::{self, File};
+use std::io::{self, Read};
 
 #[derive(Debug)]
 enum ReadUsernameError {
     IoError(io::Error),
     EmptyUsername(String),
 }
 
+impl Error for ReadUsernameError {}
```

Comment:
This code (together with the expanded imports above) make the code too large to comfortably have on a slide in a classroom. Can we take it out again?

I don't think I we show how to `impl Display` anywhere, so that's a great addition — but it's not central to the concept of error handling, so I would suggest we move it elsewhere.

Reply:
Is there some way we can hide or collapse the implementation by default in the slide view? I agree that the details of the `Display` implementation are not important, but I want to demonstrate good practices to the students, including implementing `std::error::Error` for all error types.

Reply:
> Is there some way we can hide or collapse the implementation by default in the slide view?

Partially, you can hide lines like with normal `rustdoc`: https://rust-lang.github.io/mdBook/format/mdbook.html#hiding-code-lines. However, this doesn't combine with the `editable` annotation so you have to pick your poison: an editable example or a correct one :slightly_smiling_face: 

I think it's laudable to want to be correct and show all the good practices. However, you have to keep in mind that these are _slides_. People are probably also struggling to keep everything straights at this point, so you telling them that errors should implement `Display` might make no big difference.

I'm fine either way, but my instinct would be to make the code _as small as possible_ and put all other material into the speaker notes. I'll teach the class next week, so I'll probably have more concrete feedback after that :smile:  (I'll keep the notes open on my laptop and edit them as I go... so that should let me come back with concrete experiences on what fits and what don't.)

Reply:
What I've seen happen before, and I'm pretty sure will happen here, is that people see the example on the slides as an example of a correct implementation, and copy it when they need to write something similar themselves. They might well do so some months later when they get around to writing some Rust in their project, and they've forgotten the details of what you said in the course.  With that in mind, I'd really like as much as possible for the examples we give to be correct and following best practices.

Would you be happier if I split the example out onto its own page, so you don't have to scroll?

Reply:
> Would you be happier if I split the example out onto its own page, so you don't have to scroll?

Yes, that would be great! The scrolling was something people mentioned to me this week when I taught the course. I'm thinking we need to go through all the pages again and edit them for brevity. It's an annoying battle between being complete and idiomatic while also being useful on a small screen :slightly_smiling_face:

Reply:
> Yes, that would be great!

Sorry, I was too fast... let's merge this as-is and then refactor this after the training I'm doing on Monday.